core/controllers/reader.py

```diff
@@ -1274,3 +1327,95 @@ def put(self, entity_type, entity_id):
             entity_type, state_reference,
             interaction_id, answer, answer_details)
         self.render_json({})
+
+
+class CheckpointReachedEventHandler(base.BaseHandler):
+    """Tracks a learner reaching a checkpoint."""
+
+    URL_PATH_ARGS_SCHEMAS = {
+        'exploration_id': {
+            'schema': editor.SCHEMA_FOR_EXPLORATION_ID
+        }
+    }
+    HANDLER_ARGS_SCHEMAS = {
+        'PUT': {
+            'most_recently_reached_checkpoint_exp_version': {
+                'schema': editor.SCHEMA_FOR_VERSION
+            },
+            'most_recently_reached_checkpoint_state_name': {
+                'schema': {
+                    'type': 'basestring',
+                    'validators': [{
+                        'id': 'has_length_at_most',
+                        'max_value': constants.MAX_STATE_NAME_LENGTH
+                    }]
+                }
+            },
+        }
+    }
+
+    @acl_decorators.can_play_exploration
+    def put(self, exploration_id):
+        """Handles PUT requests.
+
+        Args:
+            exploration_id: str. The ID of the exploration.
+        """
+
+        most_recently_reached_checkpoint_state_name = (
+            self.normalized_payload.get(
+                'most_recently_reached_checkpoint_state_name'))
+        most_recently_reached_checkpoint_exp_version = (
+            self.normalized_payload.get(
+                'most_recently_reached_checkpoint_exp_version'))
+
+        user_services.update_learner_checkpoint_progress(
+            self.user_id,
+            exploration_id,
+            most_recently_reached_checkpoint_state_name,
+            most_recently_reached_checkpoint_exp_version)
+
+        self.render_json(self.values)
+
+
+class ExplorationRestartEventHandler(base.BaseHandler):
+    """Tracks a learner restarting an exploration."""
+
+    URL_PATH_ARGS_SCHEMAS = {
+        'exploration_id': {
+            'schema': editor.SCHEMA_FOR_EXPLORATION_ID
+        }
+    }
+    HANDLER_ARGS_SCHEMAS = {
+        'PUT': {
+            'most_recently_reached_checkpoint_state_name': {
+                'schema': {
+                    'type': 'basestring',
+                    'validators': [{
+                        'id': 'has_length_at_most',
+                        'max_value': constants.MAX_STATE_NAME_LENGTH
+                    }]
+                },
+                'default_value': None
+            },
+        }
+    }
+
+    @acl_decorators.can_play_exploration
+    def put(self, exploration_id):
+        """Handles PUT requests.
+
+        Args:
+            exploration_id: str. The ID of the exploration.
+        """
+
+        user_id = self.user_id
```

Comment:
```suggestion

        user_id = self.user_id
```

This is not needed, right?

Reply:
Done!